File: app/src/lib/capabilities/text-to-image.ts
export interface TextToImageRequest {
  prompt: string;
  provider?: string;
  model?: string;
  width?: number;
  height?: number;
}

export interface TextToImageResult {
  bytes: Uint8Array;
  contentType: string;
  metadata: {
    provider: string;
    mode: "generate";
    prompt: string;
    model: string;
    width: number;
    height: number;
    source_url: string;
    creator: string;
    license: string;
  };
}

function cleanPrompt(prompt: string): string {
  const value = prompt.trim();
  if (!value) {
    throw new Error("prompt is required");
  }
  if (value.length > 2000) {
    throw new Error("prompt is too long");
  }
  return value;
}

async function providerRequestError(response: Response, providerLabel: string): Promise<Error> {
  const raw = (await response.text()).replace(/\s+/g, " ").trim().slice(0, 500);
  let detail = raw;
  if (raw) {
    try {
      const payload = JSON.parse(raw) as {
        error?: string | { code?: string; message?: string };
        errors?: Array<{ code?: string; message?: string }>;
        message?: string;
      };
      const firstError = payload.errors?.[0];
      if (typeof payload.error === "string") {
        detail = payload.error;
      } else if (payload.error?.message) {
        detail = payload.error.code
          ? `${payload.error.code}: ${payload.error.message}`
          : payload.error.message;
      } else if (firstError?.message) {
        detail = firstError.code ? `${firstError.code}: ${firstError.message}` : firstError.message;
      } else if (payload.message) {
        detail = payload.message;
      }
    } catch {
      // Keep the response excerpt when the provider did not return JSON.
    }
  }
  const status = `HTTP ${response.status}${response.statusText ? ` ${response.statusText}` : ""}`;
  return new Error(`${providerLabel} 请求失败（${status}）${detail ? `：${detail}` : ""}`);
}

export async function generateTextToImage(request: TextToImageRequest): Promise<TextToImageResult> {
  const provider = request.provider?.trim() || "pollinations";
  if (provider === "pollinations") {
    return generatePollinationsImage(request);
  }
  if (provider === "cloudflare_workers") {
    return generateCloudflareWorkersImage(request);
  }
  if (provider === "huggingface") {
    return generateHuggingFaceImage(request);
  }
  if (provider === "volcengine") {
    return generateVolcengineImage(request);
  }
  throw new Error(`Unsupported text-to-image provider: ${provider}`);
}

async function generatePollinationsImage(request: TextToImageRequest): Promise<TextToImageResult> {
  const prompt = cleanPrompt(request.prompt);
  const width = request.width ?? 1080;
  const height = request.height ?? 1920;
  const params = new URLSearchParams({
    width: String(width),
    height: String(height),
    nologo: "true",
    private: "true",
    safe: "true",
  });
  const resolvedModel = request.model?.trim() || process.env.KB_POLLINATIONS_IMAGE_MODEL || "";
  if (resolvedModel) {
    params.set("model", resolvedModel);
  }

  const headers: HeadersInit = {};
  const apiKey = process.env.KB_POLLINATIONS_API_KEY || process.env.POLLINATIONS_API_KEY;
  if (apiKey) {
    headers.Authorization = `Bearer ${apiKey}`;
  }

  const response = await fetch(
    `https://image.pollinations.ai/prompt/${encodeURIComponent(prompt)}?${params.toString()}`,
    { headers },
  );
  if (!response.ok) {
    throw await providerRequestError(response, "Pollinations");
  }
  const contentType = response.headers.get("content-type") ?? "";
  if (!contentType.startsWith("image/")) {
    throw new Error(`Pollinations returned non-image content: ${contentType || "unknown"}`);
  }

  return {
    bytes: new Uint8Array(await response.arrayBuffer()),
    contentType,
    metadata: {
      provider: "pollinations",
      mode: "generate",
      prompt,
      model: resolvedModel || "pollinations",
      width,
      height,
      source_url: "",
      creator: "ai_generated",
      license: "provider_terms",
    },
  };
}

async function generateCloudflareWorkersImage(request: TextToImageRequest): Promise<TextToImageResult> {
  const prompt = cleanPrompt(request.prompt);
  const width = request.width ?? 1080;
  const height = request.height ?? 1920;
  const accountId = process.env.KB_CLOUDFLARE_ACCOUNT_ID || process.env.CLOUDFLARE_ACCOUNT_ID;
  const apiToken = process.env.KB_CLOUDFLARE_API_TOKEN || process.env.CLOUDFLARE_API_TOKEN;
  if (!accountId || !apiToken) {
    throw new Error("Cloudflare Workers AI requires KB_CLOUDFLARE_ACCOUNT_ID and KB_CLOUDFLARE_API_TOKEN");
  }

  const resolvedModel = request.model?.trim()
    || process.env.KB_CLOUDFLARE_IMAGE_MODEL
    || "@cf/black-forest-labs/flux-1-schnell";
  const response = await fetch(
    `https://api.cloudflare.com/client/v4/accounts/${accountId}/ai/run/${resolvedModel}`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ prompt, width, height }),
    },
  );
  if (!response.ok) {
    throw await providerRequestError(response, "Cloudflare Workers AI");
  }

  const contentType = response.headers.get("content-type") ?? "";
  if (contentType.startsWith("application/json")) {
    const data = (await response.json()) as { result?: { image?: string }; image?: string };
    const imageBase64 = data.result?.image ?? data.image;
    if (!imageBase64) {
      throw new Error("Cloudflare Workers AI returned JSON without result.image");
    }
    const bytes = new Uint8Array(Buffer.from(imageBase64, "base64"));
    const detectedType = bytes[0] === 0xff && bytes[1] === 0xd8
      ? "image/jpeg"
      : bytes[0] === 0x89 && bytes[1] === 0x50
        ? "image/png"
        : "image/png";
    return {
      bytes,
      contentType: detectedType,
      metadata: {
        provider: "cloudflare_workers",
        mode: "generate",
        prompt,
        model: resolvedModel,
        width,
        height,
        source_url: "",
        creator: "ai_generated",
        license: "provider_terms",
      },
    };
  }
  if (!contentType.startsWith("image/")) {
    throw new Error(`Cloudflare Workers AI returned non-image content: ${contentType || "unknown"}`);
  }

  return {
    bytes: new Uint8Array(await response.arrayBuffer()),
    contentType,
    metadata: {
      provider: "cloudflare_workers",
      mode: "generate",
      prompt,
      model: resolvedModel,
      width,
      height,
      source_url: "",
      creator: "ai_generated",
      license: "provider_terms",
    },
  };
}

async function generateHuggingFaceImage(request: TextToImageRequest): Promise<TextToImageResult> {
  const prompt = cleanPrompt(request.prompt);
  const width = request.width ?? 1080;
  const height = request.height ?? 1920;
  const apiToken = process.env.KB_HUGGINGFACE_API_TOKEN
    || process.env.HUGGINGFACE_API_TOKEN
    || process.env.HF_TOKEN;
  if (!apiToken) {
    throw new Error("Hugging Face image generation requires KB_HUGGINGFACE_API_TOKEN or HF_TOKEN");
  }

  const resolvedModel = request.model?.trim()
    || process.env.KB_HUGGINGFACE_IMAGE_MODEL
    || "black-forest-labs/FLUX.1-schnell";
  const baseUrl = (process.env.KB_HUGGINGFACE_IMAGE_BASE_URL || "https://router.huggingface.co/hf-inference/models")
    .replace(/\/$/, "");
  const response = await fetch(`${baseUrl}/${resolvedModel}`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiToken}`,
      Accept: "image/png",
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      inputs: prompt,
      parameters: { width, height },
    }),
  });
  if (!response.ok) {
    throw await providerRequestError(response, "Hugging Face");
  }
  const contentType = response.headers.get("content-type") ?? "";
  if (!contentType.startsWith("image/")) {
    const detail = (await response.text()).slice(0, 300);
    throw new Error(`Hugging Face returned non-image content: ${contentType || "unknown"} ${detail}`);
  }

  return {
    bytes: new Uint8Array(await response.arrayBuffer()),
    contentType,
    metadata: {
      provider: "huggingface",
      mode: "generate",
      prompt,
      model: resolvedModel,
      width,
      height,
      source_url: "",
      creator: "ai_generated",
      license: "model_terms",
    },
  };
}

async function generateVolcengineImage(request: TextToImageRequest): Promise<TextToImageResult> {
  const prompt = cleanPrompt(request.prompt);
  const width = request.width ?? 1080;
  const height = request.height ?? 1920;
  const apiKey = process.env.KB_VOLCENGINE_IMAGE_API_KEY
    || process.env.KB_VOLC_IMAGE_API_KEY
    || process.env.ARK_API_KEY;
  if (!apiKey) {
    throw new Error(
      "Volcengine image generation requires KB_VOLCENGINE_IMAGE_API_KEY, KB_VOLC_IMAGE_API_KEY, or ARK_API_KEY",
    );
  }

  const resolvedModel = request.model?.trim()
    || process.env.KB_VOLCENGINE_IMAGE_MODEL
    || "doubao-seedream-4-0-250828";
  const resolvedSize = process.env.KB_VOLCENGINE_IMAGE_SIZE?.trim() || "2K";
  const baseUrl = (process.env.KB_VOLCENGINE_IMAGE_BASE_URL || "https://ark.cn-beijing.volces.com/api/v3")
    .replace(/\/$/, "");
  const endpoint = baseUrl.endsWith("/images/generations")
    ? baseUrl
    : `${baseUrl}/images/generations`;
  const response = await fetch(endpoint, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: resolvedModel,
      prompt,
      sequential_image_generation: "disabled",
      response_format: "url",
      size: resolvedSize,
      stream: false,
      watermark: false,
    }),
  });
  if (!response.ok) {
    throw await providerRequestError(response, "火山引擎方舟");
  }

  const payload = (await response.json()) as {
    data?: Array<{ url?: string; b64_json?: string }>;
  };
  const item = payload.data?.[0];
  if (!item) {
    throw new Error("Volcengine returned JSON without data[0]");
  }

  let bytes: Uint8Array;
  let contentType: string;
  if (item.url) {
    const imageResponse = await fetch(item.url);
    if (!imageResponse.ok) {
      throw new Error(`Volcengine image download failed: ${imageResponse.status}`);
    }
    bytes = new Uint8Array(await imageResponse.arrayBuffer());
    contentType = imageResponse.headers.get("content-type") ?? "";
  } else if (item.b64_json) {
    bytes = new Uint8Array(Buffer.from(item.b64_json, "base64"));
    contentType = "image/png";
  } else {
    throw new Error("Volcengine returned data[0] without url or b64_json");
  }

  const looksLikeImage = contentType.startsWith("image/")
    || (bytes[0] === 0xff && bytes[1] === 0xd8)
    || (bytes[0] === 0x89 && bytes[1] === 0x50)
    || (bytes[0] === 0x52 && bytes[1] === 0x49 && bytes[2] === 0x46 && bytes[3] === 0x46);
  if (!looksLikeImage) {
    throw new Error(`Volcengine returned non-image content: ${contentType || "unknown"}`);
  }
  if (!contentType.startsWith("image/")) {
    contentType = bytes[0] === 0xff ? "image/jpeg" : bytes[0] === 0x52 ? "image/webp" : "image/png";
  }

  return {
    bytes,
    contentType,
    metadata: {
      provider: "volcengine",
      mode: "generate",
      prompt,
      model: resolvedModel,
      width,
      height,
      source_url: "",
      creator: "ai_generated",
      license: "provider_terms",
    },
  };
}

File: knowbreak/capabilities/text_to_image.py
"""Text-to-image capability.

This module owns provider-specific generation details and returns bytes plus
metadata. Pipeline stages decide where to write files and how to reference them.
"""

from __future__ import annotations

import base64
from dataclasses import dataclass
from urllib.parse import quote

import httpx

from ..config import Config

POLLINATIONS_IMAGE = "https://image.pollinations.ai/prompt"


@dataclass(frozen=True)
class GeneratedImage:
    content: bytes
    metadata: dict


def generate_text_to_image(
    cfg: Config,
    prompt: str,
    *,
    provider: str = "pollinations",
    model: str | None = None,
    width: int = 1080,
    height: int = 1920,
) -> GeneratedImage:
    if provider == "pollinations":
        return _generate_pollinations(cfg, prompt, model=model, width=width, height=height)
    if provider == "cloudflare_workers":
        return _generate_cloudflare_workers(cfg, prompt, model=model, width=width, height=height)
    if provider == "huggingface":
        return _generate_huggingface(cfg, prompt, model=model, width=width, height=height)
    if provider == "volcengine":
        return _generate_volcengine(cfg, prompt, model=model, width=width, height=height)
    raise ValueError(f"Unsupported text-to-image provider: {provider}")


def _generate_pollinations(
    cfg: Config,
    prompt: str,
    *,
    model: str | None,
    width: int,
    height: int,
) -> GeneratedImage:
    params = {
        "width": str(width),
        "height": str(height),
        "nologo": "true",
        "private": "true",
        "safe": "true",
    }
    resolved_model = model or cfg.pollinations_image_model
    if resolved_model:
        params["model"] = resolved_model
    headers = {"Authorization": f"Bearer {cfg.pollinations_api_key}"} if cfg.pollinations_api_key else None
    url = f"{POLLINATIONS_IMAGE}/{quote(prompt)}"
    r = httpx.get(url, params=params, headers=headers, timeout=90, follow_redirects=True)
    r.raise_for_status()
    content_type = r.headers.get("content-type", "")
    if not content_type.startswith("image/"):
        raise ValueError(f"Pollinations returned non-image content: {content_type}")
    return GeneratedImage(
        content=r.content,
        metadata={
            "mode": "generate",
            "source_url": "",
            "creator": "ai_generated",
            "license": "provider_terms",
            "prompt": prompt,
            "model": resolved_model or "pollinations",
            "width": width,
            "height": height,
        },
    )


def _generate_cloudflare_workers(
    cfg: Config,
    prompt: str,
    *,
    model: str | None,
    width: int,
    height: int,
) -> GeneratedImage:
    if not cfg.cloudflare_account_id or not cfg.cloudflare_api_token:
        raise ValueError("Cloudflare Workers AI requires KB_CLOUDFLARE_ACCOUNT_ID and KB_CLOUDFLARE_API_TOKEN")
    resolved_model = model or cfg.cloudflare_image_model
    url = (
        "https://api.cloudflare.com/client/v4/accounts/"
        f"{cfg.cloudflare_account_id}/ai/run/{resolved_model}"
    )
    r = httpx.post(
        url,
        headers={"Authorization": f"Bearer {cfg.cloudflare_api_token}"},
        json={"prompt": prompt, "width": width, "height": height},
        timeout=90,
    )
    r.raise_for_status()
    content_type = r.headers.get("content-type", "")
    if content_type.startswith("application/json"):
        data = r.json()
        image_b64 = data.get("result", {}).get("image") or data.get("image")
        if not image_b64:
            raise ValueError("Cloudflare Workers AI returned JSON without result.image")
        import base64

        content = base64.b64decode(image_b64)
        content_type = "image/png"
    else:
        content = r.content
    if not content_type.startswith("image/"):
        raise ValueError(f"Cloudflare Workers AI returned non-image content: {content_type}")
    return GeneratedImage(
        content=content,
        metadata={
            "mode": "generate",
            "source_url": "",
            "creator": "ai_generated",
            "license": "provider_terms",
            "prompt": prompt,
            "model": resolved_model,
            "width": width,
            "height": height,
        },
    )


def _generate_huggingface(
    cfg: Config,
    prompt: str,
    *,
    model: str | None,
    width: int,
    height: int,
) -> GeneratedImage:
    if not cfg.huggingface_api_token:
        raise ValueError("Hugging Face image generation requires KB_HUGGINGFACE_API_TOKEN or HF_TOKEN")
    resolved_model = model or cfg.huggingface_image_model
    base_url = cfg.huggingface_image_base_url.rstrip("/")
    url = f"{base_url}/{resolved_model}"
    r = httpx.post(
        url,
        headers={
            "Authorization": f"Bearer {cfg.huggingface_api_token}",
            "Accept": "image/png",
        },
        json={
            "inputs": prompt,
            "parameters": {
                "width": width,
                "height": height,
            },
        },
        timeout=120,
    )
    if r.status_code != 200:
        err_msg = f"Hugging Face image generation failed: {r.status_code}"
        try:
            payload = r.json()
            if isinstance(payload, dict) and "error" in payload:
                err_msg = f"Hugging Face error ({r.status_code}): {payload['error']}"
        except Exception:
            pass
        raise RuntimeError(err_msg)
    content_type = r.headers.get("content-type", "")
    if not content_type.startswith("image/"):
        detail = r.text[:300]
        raise ValueError(f"Hugging Face returned non-image content: {content_type} {detail}")
    return GeneratedImage(
        content=r.content,
        metadata={
            "mode": "generate",
            "source_url": "",
            "creator": "ai_generated",
            "license": "model_terms",
            "prompt": prompt,
            "model": resolved_model,
            "width": width,
            "height": height,
        },
    )


def _generate_volcengine(
    cfg: Config,
    prompt: str,
    *,
    model: str | None,
    width: int,
    height: int,
) -> GeneratedImage:
    """Generate an image through Volcengine Ark's Images API."""
    if not cfg.volcengine_image_api_key:
        raise ValueError(
            "Volcengine image generation requires KB_VOLCENGINE_IMAGE_API_KEY, "
            "KB_VOLC_IMAGE_API_KEY, or ARK_API_KEY"
        )
    resolved_model = model or cfg.volcengine_image_model
    if not resolved_model:
        raise ValueError("Volcengine image generation requires KB_VOLCENGINE_IMAGE_MODEL")

    base_url = cfg.volcengine_image_base_url.rstrip("/")
    url = base_url if base_url.endswith("/images/generations") else f"{base_url}/images/generations"
    r = httpx.post(
        url,
        headers={
            "Authorization": f"Bearer {cfg.volcengine_image_api_key}",
            "Content-Type": "application/json",
        },
        json={
            "model": resolved_model,
            "prompt": prompt,
            "sequential_image_generation": "disabled",
            "response_format": "url",
            "size": cfg.volcengine_image_size,
            "stream": False,
            "watermark": False,
        },
        timeout=180,
    )
    if r.status_code != 200:
        detail = r.text[:500]
        try:
            payload = r.json()
            if isinstance(payload, dict):
                error = payload.get("error")
                if isinstance(error, dict):
                    message = str(error.get("message") or error)
                    code = error.get("code")
                    detail = f"{code}: {message}" if code else message
                elif error:
                    detail = str(error)
        except Exception:
            pass
        raise RuntimeError(f"Volcengine image generation failed ({r.status_code}): {detail}")

    payload = r.json()
    data = payload.get("data") if isinstance(payload, dict) else None
    item = data[0] if isinstance(data, list) and data and isinstance(data[0], dict) else None
    if not item:
        raise ValueError("Volcengine returned JSON without data[0]")

    image_url = item.get("url")
    image_base64 = item.get("b64_json")
    if isinstance(image_url, str) and image_url:
        image_response = httpx.get(image_url, timeout=180, follow_redirects=True)
        image_response.raise_for_status()
        content = image_response.content
        content_type = image_response.headers.get("content-type", "")
    elif isinstance(image_base64, str) and image_base64:
        content = base64.b64decode(image_base64)
        content_type = "image/png"
    else:
        raise ValueError("Volcengine returned data[0] without url or b64_json")

    if not content or not (
        content_type.startswith("image/")
        or content.startswith(b"\xff\xd8")
        or content.startswith(b"\x89PNG")
        or content.startswith(b"RIFF")
    ):
        raise ValueError(f"Volcengine returned non-image content: {content_type or 'unknown'}")

    return GeneratedImage(
        content=content,
        metadata={
            "mode": "generate",
            "source_url": "",
            "creator": "ai_generated",
            "license": "provider_terms",
            "prompt": prompt,
            "model": resolved_model,
            "width": width,
            "height": height,
        },
    )
